### PR TITLE
Stack evolves, use lucky egg to maximize XP

### PR DESCRIPTION
**Pull request by svombergen**
### Prerequisites
Make sure you remove the pokemons you want to keep for evolving from the ObligatoryTransferList and add them to the IgnoredPokemonList

**Changes made: 	**

 1. Facilitate evolve stacking, configurable with 2 variables: 
- evolveBeforeTransfer - list of pokemon to save up candies and pokemon for evolve stacking
- evolveStackLimit - threshold to pop lucky egg and evolve all saved up evolves at once

 2. Made logging spam of walkingLoop cycle dependent on displayKeepalive setting.

** Changes detail**

- Added 2 variables in settings and template file, evolveBeforeTransfer and evolveStackLimit
- Added EvolvePokemon task to handle Evolve stacking logic.
- Added call to task EvolvePokemon in Bot.kt ProfileLoop
- Added 'if (settings.displayKeepalive)' statement to logging message in WalkingLoop of Walk.kt to avoid spam in console.

### CODE REVIEW

**Comment by zhoushifeng8** [comment]
when can use this evolve function?

**Comment by svombergen** [comment]
When jabbink decides to merge it.

**Comment by svombergen** [comment]
I expect the release logic could use a good refactoring, after adding evolves. The logic for releasing will be divided and intertwined on 3 places.

File: src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                // Get pokemonFamily through this val (can get this somewhere else?):
+                val lastPokemonInGroup = it.value.last()
+                var maxPossibleEvolves = bot.api.inventories.candyjar.getCandies(lastPokemonInGroup.pokemonFamily) / lastPokemonInGroup.candiesToEvolve
+                // Add the minimum value, depending on which is the bottleneck, amount of candy, or pokemon of this type in pokebank:
+                countEvolveStack +=  Math.min(maxPossibleEvolves,it.value.count())
+                // Release pokemon we wont be able to evolve anyway, because of lack of candy:
+                it.value.sortedByDescending { it.cp }.forEach {
+                    maxPossibleEvolves--
+                    if (maxPossibleEvolves < 0) {
+                        releasePokemon(it,"Not enough candy to save for evolve",ctx)
```
Comment (jabbink) [inline src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt]: releasing a pokemon increases your candy count so you could update the maxPossibleEvolves again :smile:

**Comment by jabbink** [comment]
Can you please use the default IntelliJ autoformat eg `if(settings` becomes `if (settings`

Also how does this work with transferring pokemon based on the cp/iv thresholds and maxPokemonAmount/keepPokemonAmount?

File: src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+class EvolvePokemon : Task {
+    override fun run(bot: Bot, ctx: Context, settings: Settings) {
+        //count the current stack of possible evolves
+        var countEvolveStack = 0
+        val groupedPokemonForCount = ctx.api.inventories.pokebank.pokemons.groupBy { it.pokemonId }
+        groupedPokemonForCount.forEach {
+            if(settings.evolveBeforeTransfer.contains(it.key)){
+                // Get pokemonFamily through this val (can get this somewhere else?):
+                val lastPokemonInGroup = it.value.last()
+                var maxPossibleEvolves = bot.api.inventories.candyjar.getCandies(lastPokemonInGroup.pokemonFamily) / lastPokemonInGroup.candiesToEvolve
```
Comment (jabbink) [inline src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt]:
```kotlin

                val pokemonMeta = PokemonMetaRegistry.getMeta(it.key)
                var maxPossibleEvolves = bot.api.inventories.candyjar.getCandies(pokemonMeta.family) / pokemonMeta.candyToEvolve
```

File: src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt (hunk 1 above)
Comment (svombergen) [inline src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt]: oh you are right, my lazy code will just check next profileloop rotation :smile:

File: src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt (hunk 2 above)
Comment (svombergen) [inline src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt]: ah thank you, will update

**Comment by svombergen** [comment]
Prerequisite is putting the pokemons you want to evolve into the ignoredPokemon list (and remove from obligatory). The lists are getting pretty redundant, so evolve and release tasks could use a good refactoring.

**Comment by jabbink** [comment]
>so evolve and release tasks could use a good refactoring.

agreed

**Comment by jabbink** [comment]
Will test now

File: src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        val groupedPokemonForCount = ctx.api.inventories.pokebank.pokemons.groupBy { it.pokemonId }
+        groupedPokemonForCount.forEach {
+            if (settings.evolveBeforeTransfer.contains(it.key)){
+                // Get pokemonFamily meta information
+                val pokemonMeta = PokemonMetaRegistry.getMeta(it.key)
+                var maxPossibleEvolves = bot.api.inventories.candyjar.getCandies(pokemonMeta.family) / pokemonMeta.candyToEvolve
+                // Add the minimum value, depending on which is the bottleneck, amount of candy, or pokemon of this type in pokebank:
+                countEvolveStack +=  Math.min(maxPossibleEvolves,it.value.count())
+                // Release pokemon we wont be able to evolve anyway, because of lack of candy:
+                it.value.sortedByDescending { it.cp }.forEach {
```
Comment (devinsba) [inline src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt]: Shouldn't this use IV if their settings have IV selected as their sort?

File: src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+            }
+        }
+        Log.yellow("Stack of pokemon ready to evolve: $countEvolveStack pokemons")
+
+        // use lucky egg if above evolve stack limit and evolve the whole stack
+        if (countEvolveStack >= settings.evolveStackLimit) {
+            val startingXP = ctx.api.playerProfile.stats.experience
+            Log.yellow("Using Lucky Egg before evolving stack of $countEvolveStack pokemons")
+            var resultLuckyEgg = ctx.api.inventories.itemBag.useLuckyEgg()
```
Comment (jabbink) [inline src/main/kotlin/ink/abb/pogo/scraper/tasks/EvolvePokemon.kt]: please use `ctx.api.cachedInventories` and wrap in try/catch because it may fail

**Comment by jabbink** [comment]
Looks good, please see line comments and auto indentation issue.

**Comment by svombergen** [comment]
Just merged develop branch and:
- Added try catch around lucky egg usage, also changed to cachedInventory
- Added sorting by Iv when releasing
- Corrected indentation

**Comment by jabbink** [comment]
>Prerequisite is putting the pokemons you want to evolve into the ignoredPokemon list (and remove from obligatory).

Maybe to prevent confusion when this gets merged, change the shouldTransfer function(s?) to early return when the pokemon is in the evolveBeforeTransfer list

**Comment by jabbink** [comment]
also about overhaul #860

**Comment by svombergen** [comment]
Agreed, will do some refactoring to prevent confusion later today.

**Comment by yuanweilin** [comment]
add the 
"evolve_before_transfer" : [ "DODUO", "RATTATA", "CATERPIE", "PIDGEY", "ZUBAT", "GOLDEEN", "MAGMAR", "PINSIR", "GOLBAT", "WEEDLE" ],
 "evolve_stack_limit" : 100,
in my setting
when i run java -jar PokemonGoBot-0.5.2.jar 
It will be deleted

**Comment by svombergen** [comment]
because it is not merged into 0.5.2, just a little more patience :)

**Comment by yuanweilin** [comment]
lol,  I'm so sad

**Comment by svombergen** [comment]
Alright committed shouldrelease logic to avoid confusion. Will truely have a look at refactoring all logic around evolving and releasing when i get the time.

**Comment by gise88** [comment]
I know.. I am in late because this was already merged but I am going to ask if you can watch at this: https://github.com/jabbink/PokemonGoBot/issues/1040

Thanks anyway! :smiley:

**Comment by Sieberkev** [comment]
Halfway the evolving (50/100 pokes) I got:

```
14 aug 13:58:59 [default: ProfileLoop] - Error running loop ProfileLoop!
com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred.
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
        at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
        at com.pokegoapi.api.pokemon.Pokemon.evolve(Pokemon.java:231)
        at ink.abb.pogo.scraper.tasks.EvolvePokemon.run(EvolvePokemon.kt:69)
        at ink.abb.pogo.scraper.Bot.task(Bot.kt:254)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:142)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:35)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:202)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:35)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at rx.exceptions.Exceptions.propagate(Exceptions.java:58)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:465)
        at rx.observables.BlockingObservable.first(BlockingObservable.java:168)
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:37)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:109)
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:86)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:74)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:43)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:9957)
        at rx.Observable.subscribe(Observable.java:9924)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445)
        ... 11 more
Caused by: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.ServerRequest.getData(ServerRequest.java:78)
        at com.pokegoapi.main.RequestHandler.run(RequestHandler.java:296)
        at java.lang.Thread.run(Thread.java:745)

```

Maybe the evolving is too fast?

**Comment by johnnyslt** [comment]
Same thing for me, while using the bot from source:

`      14 Aug 22:26:31 [default: ProfileLoop] - Going to transfer RATICATE with CP 749 and IV 53%; reason: Evolved for XP only
       14 Aug 22:26:32 [default: ProfileLoop] - Evolving PIDGEY before transfer
       14 Aug 22:26:32 [default: ProfileLoop] - Error running loop ProfileLoop!
com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred.
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
        at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
        at com.pokegoapi.api.pokemon.Pokemon.evolve(Pokemon.java:231)
        at ink.abb.pogo.scraper.tasks.EvolvePokemon.run(EvolvePokemon.kt:69)
        at ink.abb.pogo.scraper.Bot.task(Bot.kt:254)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:142)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:35)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:202)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:35)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at rx.exceptions.Exceptions.propagate(Exceptions.java:58)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:465)
        at rx.observables.BlockingObservable.first(BlockingObservable.java:168)
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:37)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:109)
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:86)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:74)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:43)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:9957)
        at rx.Observable.subscribe(Observable.java:9924)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445)
        ... 11 more
Caused by: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.ServerRequest.getData(ServerRequest.java:78)
        at com.pokegoapi.main.RequestHandler.run(RequestHandler.java:296)
        at java.lang.Thread.run(Thread.java:745)
`

**Comment by johnnyslt** [comment]
I get the same thing:

>       14 Aug 22:26:31 [default: ProfileLoop] - Going to transfer RATICATE with CP 749 and IV 53%; reason: Evolved for XP only
       14 Aug 22:26:32 [default: ProfileLoop] - Evolving PIDGEY before transfer
       14 Aug 22:26:32 [default: ProfileLoop] - Error running loop ProfileLoop!
com.pokegoapi.exceptions.AsyncPokemonGoException: Unknown exception occurred.
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:45)
        at com.pokegoapi.main.RequestHandler.sendServerRequests(RequestHandler.java:155)
        at com.pokegoapi.api.pokemon.Pokemon.evolve(Pokemon.java:231)
        at ink.abb.pogo.scraper.tasks.EvolvePokemon.run(EvolvePokemon.kt:69)
        at ink.abb.pogo.scraper.Bot.task(Bot.kt:254)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:142)
        at ink.abb.pogo.scraper.Bot$start$3.invoke(Bot.kt:35)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:202)
        at ink.abb.pogo.scraper.Bot$runLoop$1.invoke(Bot.kt:35)
        at kotlin.concurrent.ThreadsKt$thread$thread$1.run(Thread.kt:18)
Caused by: java.lang.RuntimeException: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at rx.exceptions.Exceptions.propagate(Exceptions.java:58)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:465)
        at rx.observables.BlockingObservable.first(BlockingObservable.java:168)
        at com.pokegoapi.util.AsyncHelper.toBlocking(AsyncHelper.java:37)
        ... 9 more
Caused by: java.util.concurrent.ExecutionException: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:109)
        at com.pokegoapi.main.RequestHandler$1.get(RequestHandler.java:86)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:74)
        at rx.internal.operators.OnSubscribeToObservableFuture$ToObservableFuture.call(OnSubscribeToObservableFuture.java:43)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
        at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
        at rx.Observable.subscribe(Observable.java:9957)
        at rx.Observable.subscribe(Observable.java:9924)
        at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:445)
        ... 11 more
Caused by: com.google.protobuf.InvalidProtocolBufferException: Contents of buffer are null
        at com.pokegoapi.main.ServerRequest.getData(ServerRequest.java:78)
        at com.pokegoapi.main.RequestHandler.run(RequestHandler.java:296)
        at java.lang.Thread.run(Thread.java:745)

**Pull request merged**